### PR TITLE
Add tap reporter

### PR DESCRIPTION
**Pull request by TheAlphaNerd**
This is a first pass at adding a tap-reporter to citgm. The reporter is exposes through a command line flag `-t <filepath>`, where if the optional file path is not provided it simply logs directly to STDOUT.

In order to support this reporter there has been a refactor of the general post process reporting that we have done. Outputting the results via the default logger, markdown, or tap are now all sharing the same basic abstraction for the most part.

This has yet to be tested in CI, but it should in theory allow us to start looking at tap output for a basic idea of the results of a run.

Other than the basic error propagated by citgm there is no additional information currently included. I think this is something we can try to improve in a future iteration, if we can find a better way of extrapolating error information from tests run in child processes.

### CODE REVIEW

**Comment by TheAlphaNerd** [comment]
potentially closes #85

**Comment by rvagg** [comment]
I don't have time for an in-depth look at this but on a quick scan it _seems_ OK, I'd like to see output from it, however, can we run it through Jenkins?

**Comment by TheAlphaNerd** [comment]
@rvagg just trying to set it up to run on jenkins right now

**Comment by TheAlphaNerd** [comment]
Here's a citgm run with the new option enabled. I have simply dumped the tap output to a file that exists at `$PWD/test.tap`

I'm pretty sure that is where jenkins is going to look for it... so fingers crossed :smile:

File: .eslintrc [hunk 1]
```diff
@@ -18,10 +18,11 @@
         ]
     },
     "env": {
-        "node": true
+        "node": true,
```
Comment (thefourtheye) [inline .eslintrc]: This is not well formed json

Comment (TheAlphaNerd) [inline .eslintrc]: good eye... just cleaned that up

**Comment by TheAlphaNerd** [comment]
Looks like it is working! I found a couple small regressions I introduced.

After writing a few more tests and fixing things I'm running it again --> https://ci.nodejs.org/job/thealphanerd-tap-smoker/6/

This time against v4.4.0-rc.2

**Comment by jasnell** [comment]
Nice! LGTM

File: lib/reporter/markdown.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  if (!(modules instanceof Array)) {
+    modules = [modules];
+  }
+
+  var passed = util.getPassing(modules);
+  var flaky = util.getFlakyFails(modules);
+  var failed = util.getFails(modules);
+
+  if (failed.length === 0) {
+    logger('## 🎉🎉 CITGM Passed 🎉🎉');
```
Comment (thefourtheye) [inline lib/reporter/markdown.js]: What are these symbols?

File: bin/citgm-all [hunk 3]
```diff
@@ -102,93 +101,39 @@ function runCitgm (mod, name, next) {
...
     if (app.markdown) {
-      printMarkdown();
+      reporter.markdown(log.bypass, modules);
+    }
+    
+    if (app.tap) {
+      // if tap is a string it should be a path to write output to
+      // if not use `log.bypass` which is currently process.stdout.write
+      // TODO check that we can write to that path, perhaps require a flag to overwrite
+      let tap = (typeof app.tap === 'string') ? app.tap : log.bypass;
```
Comment (thefourtheye) [inline bin/citgm-all]: Why can't `bypass` be in `reporter`?

File: lib/reporter/logger.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+var _ = require('lodash');
+var chalk = require('chalk');
+
+var util = require('./util');
+
+function logModule(log, logType, module) {
+  log[logType](chalk.yellow('module name:'), module.name);
+  log[logType](chalk.yellow('version:'), module.version);
+  if (module.error) {
+    log[logType]('error', module.error.message);
```
Comment (thefourtheye) [inline lib/reporter/logger.js]: Shouldn't this be in red color?

File: lib/reporter/util.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+}
+
+function getFails(modules) {
+  return _.filter(modules, function (module) {
+    return (module.error && !module.flaky);
+  });
+}
+
+function hasFailures(modules) {
+  return getFails(modules).length >= 1;
```
Comment (thefourtheye) [inline lib/reporter/util.js]: I would prefer `_.some` here.

File: lib/reporter/markdown.js (hunk 2 above)
Comment (TheAlphaNerd) [inline lib/reporter/markdown.js]: Fun emoji for the markdown reporter. This was [discussed when they were originally merged](https://github.com/nodejs/citgm/pull/40)

More than happy to discuss again if you think they could cause an issue

File: bin/citgm-all (hunk 3 above)
Comment (TheAlphaNerd) [inline bin/citgm-all]: The idea here was that the reporter simply gets a function that has the signature `fn(msg)` which makes it simple to stub in different reporter options. I'm also not 100% that "bypass" is going to survive so it was a quick way to isolate

File: lib/reporter/logger.js (hunk 4 above)
Comment (TheAlphaNerd) [inline lib/reporter/logger.js]: The logger already wraps the title for an error in red
https://github.com/TheAlphaNerd/citgm/blob/reporter/lib/out.js#L65-L67
this is simply passing the error.message

File: lib/reporter/util.js (hunk 5 above)
Comment (TheAlphaNerd) [inline lib/reporter/util.js]: udpated

**Comment by thefourtheye** [comment]
LGTM. The failures could show better error messages, but then that can wait for another PR.

**Comment by TheAlphaNerd** [comment]
@thefourtheye if you have any ideas on how to improve failures I'd be all ears. It will likely require a decent refactor to the way we are logging right now. All the post reporting is doing so with an object that has been prepared in memory while the job is going on. We would likely need to move to a stream based interface for all logging if we hope to have more granular tap output

**Comment by TheAlphaNerd** [comment]
So I just took a quick pass at it and I think I have a naive solution to getting reporting in with tap results. I am interested to see if this causes any memory problems when run with citgm-all... Fingers crossed.

**Comment by TheAlphaNerd** [comment]
New citgm => https://ci.nodejs.org/job/thealphanerd-tap-smoker/8/

**Comment by TheAlphaNerd** [comment]
There was some weirdness with ansi scripts that were using the `\r` character

I've attempted to introduce a new option to strip ansi, it is opt in on a per module basis
==> https://ci.nodejs.org/job/thealphanerd-tap-smoker/12/

**Comment by gibm** [comment]
I noticed that there are a lot of blank lines in the output for some modules (e.g. express) in [the latest CI runs](https://ci.nodejs.org/job/thealphanerd-tap-smoker/13/nodes=ppcle-ubuntu1404/tapResults/), not sure if that's related to the ansi weirdness you mentioned.

**Comment by TheAlphaNerd** [comment]
@gibm Those lines are unrelated to the weirdness I was originally seeing. A number of the different test runners have output that normally would generate animations in the terminal during test running. When logging that information it ends up looking like the express output in that example

**Comment by gibm** [comment]
@TheAlphaNerd that makes sense. Is there any easy way to remove them? Possibly something for a later revision?

**Comment by TheAlphaNerd** [comment]
@gibm there is an output sanitization step... we should be able to modify that and be more aggressive in what we clean up

**Pull request merged**